### DISCUSSION TITLE
For loop in mouse Move event.

### DISCUSSION THREAD

**Discussion by Ted-juh** [RomanKubiak/ctrlr#266] [General]
In a CustomComponent, in the Mouse Move function, I can do While loops, Repeat loops, but a For loop doesn't run in some way. If I break the "For" loop into if/ else statements, it works as expected. 

But is there any reason why this doesn't run:
```
for i = 1,16 do
 if  _G["arp"..i]:getByte(0) == noteY then 
note = i
else 
note = 128
end
```

But this does?
```
if _G["arp1"]:getByte(0) == noteY then note = 1
elseif _G["arp2"]:getByte(0) == noteY then note = 2
-------
elseif _G["arp16"]:getByte() == noteY then note = 16
else note = 128
end
```

Ok, using a while-loop does the trick as well:
```
note = 1
while _G["arp"..note]:getByte(0) ~= noteY do note = note + 1
if note > 16 then note = 0x80 break end
end
```

But those while and repeat loops are killing me if I write something that incidentally loops forever, which forces me to close Ctrlr and I forgot to save.

**Comment by dobo365** [comment]
I personally don't think it is a good idea to put while loops in a MouseMove function, knowing that the mouse is constantly moving...
I don't know what your plans are but maybe that the uiXYSurface component could help you there.

**Reply by Ted-juh** [reply]
My plans? Don't get me started, lol.

I get what you are saying. Because the mouse is constantly moving, those loops are running constantly as well. I'm aware of that. But CPU usage seems to stay below 3%, doesn't spike. I guess that the time that the MouseMove function needs to run is below the refresh/ update time of the MouseMove event. 

But my plan is: Update a MemoryBlock with certain coordinates. But not the original event.x or event.y coordinates. There is some math done on those events that translate them into the format that I need them. I.e. event.x gets translated to (step1 - step 32) and event.y gets translated to (0-127).  It holds all the information (that's in the black boxes underneath the arp window, see pic) that's needed in the other functions and gets changed "on the fly". 

With those translated coordinates I can change the correct MemoryBlock (out of 16 MemoryBlocks) that need to be altered. Those 16 MemoryBlocks hold the sysex to be sent to the synth for the arpeggiator. Those MemoryBlocks are also used to paint the blocks in the component. But also to set the correct mouse cursor. And since I have to do all the math to get the right mouse cursor anyway, why duplicate all the math to the MouseDown function while it's already been taken care of? (Or to avoid a lot of global variables, it's now just one global MemoryBlock.)

![example arc](https://user-images.githubusercontent.com/71118185/115991786-6451ee80-a5ca-11eb-8b1f-6f23a6db3fa3.PNG)

Now when I hover over the first red block at F#4, it shows a mouse cursor that points to all directions. I can drag it up and down to set the velocity, I can drag it to the right to make the block larger. Let's make the block 4 steps wide. When I release it and hover over it again, it only needs to show the left/ right drag option when I'm at the last block (step 4 in this case). And the up/ down drag option anywhere but the last block (step 1 - 3).

This cannot be done at mouseDown but in my option only at MouseMove, right? That's why I need those loops. Whether it's a wise thing to do or not.  Or is there another way that I'm not aware of?

**Reply by Godlike-Productions** [reply]
There are some methods for focus that might work better for you. (I think there's something in one of the Keyboard Events demo panels), but they would probably require the blocks to be an object or component or modulator of some sort.

Instead of a while loop, can't you just put it within a mouseMoved callback. Let the callback essentially do the while loop bit.

I'm just shooting from the hip here. I don't know custom components all that well, but this looks awesome.

**Reply by Ted-juh** [reply]
> There are some methods for focus that might work better for you. (I think there's something in one of the Keyboard Events demo panels), but they would probably require the blocks to be an object or component or modulator of some sort.
> 

I guess you mean "originatingComponent". You are right they require the blocks to be a component within the (custom) component. But I  do not have the slightest idea how to add them to the custom component as a component. That deserves a new question on itself to discuss. 

>Instead of a while loop, can't you just put it within a mouseMoved callback. Let the callback essentially do the while loop bit.
>

It's already in a MouseMove function. I needed to iterate over a few MemoryBlocks within that function. It's the "for" loop that doesn't want to run. The while and repeat loops do run. I got it working in the meantime with while loops though. I just thought it was strange that the for loops do not do their job, hence the question.

>but this looks awesome.
>
Wait until you see it scroll up and down. Mind that the oct/ step buttons and the slider are all part of the custom component.  :)
